Reimplement `Wallet`, `ElectrumExt` and `Esplora{Async}Ext` with redesigned structures.

evanlinjin opened the pull request:
### Description

Closes #938

* Update `Wallet` to use redesigned structures.
* Update `bdk_electrum::ElectrumExt` to produce updates for redesigned structures.
* Update `bdk_esplora::EsploraExt` and `bdk_esplora::EsploraAsyncExt` to produce updates for redesigned structures.
* Added `example-crates/example_cli` library for implementing examples with redesigned structures.
* Added `example-crate/example_electrum` which is an electrum CLI wallet using the redesigned structures.
* Updated `example-crate/{wallet_electrum|wallet_esplora|wallet_esplora_async}` examples to use redesigned structures.
* Remove all old structures.

### Notes to the reviewers

~These changes bump our `all-features` MSRV to `1.60.0` because of the introduction of `bdk_esplora`. As long as the `bdk_chain` and `bdk_wallet` crates hit a MSRV of `1.48.0`, it will be fine (this work is done in #987).~ No longer needed due to #993 

~I had to comment out the examples that use `Wallet` with our chain sources. Once we update the helper-packages for those chain sources, we can also update the examples.~

Possible future improvements for `ElectrumExt`:

* Remove requirement to retry obtaining ALL data after reorg is detected. Transactions can be anchored to a lower block (not block tip), and an `assume_final_depth` value can be used.

* The logic to finalize an update with confirmation time can be improved during reorgs to not require returning an error.

* Use the subscription model of electrum, as intended by the API.

### Changelog notice

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

Shourya742 committed:
Rename fields of `tx_graph::Additions`

* Changed `tx` to `txs`
* Changed `txout` to `txouts`

evanlinjin committed:
[wallet_redesign] Update `Wallet` with redesigned structures

evanlinjin committed:
[wallet_redesign] Move the majority of `Update` to `bdk_chain`

This is to make it easier for chain source crates to formulate updates.

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -1685,52 +1758,62 @@ impl<D> Wallet<D> {
+
+        let changed = !changeset.is_empty();
+        if changed {
+            self.persist.stage(changeset);
```
No need for this branch. If it is empty it'll just be a noop.

evanlinjin replied:
Removed all unnecessary branches.

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -1685,52 +1758,62 @@ impl<D> Wallet<D> {
     /// Get a reference to the inner [`TxGraph`](bdk_chain::tx_graph::TxGraph).
-    pub fn as_graph(&self) -> &bdk_chain::tx_graph::TxGraph {
-        self.keychain_tracker.graph()
+    pub fn as_graph(&self) -> &TxGraph<ConfirmationTimeAnchor> {
```
```suggestion
    pub fn tx_graph(&self) -> &TxGraph<ConfirmationTimeAnchor> {
```

I think `as_` doesn't feel right. `Wallet` is not wrapping a `TxGraph` it just has one you can access. Also it's nice to have `tx_` in front of `graph` since it's not clear what it's a graph of.

evanlinjin replied:
Changed.

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -497,41 +528,80 @@ impl<D> Wallet<D> {
-        let changeset = self.keychain_tracker.insert_tx(tx, position)?;
-        let changed = changeset.is_empty();
-        self.persist.stage(changeset);
+        seen_at: Option<u64>,
```
surely you can either insert a tx in a conirmed position OR with a seent_at but not both. You should specify the `seen_at` at the Unconfirmed branch of `ConfirmationTime` or something.

evanlinjin replied:
Updated to have the `ConfirmationTime::Unconfirmed` variant have a `last_seen` field.

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -472,10 +498,15 @@ impl<D> Wallet<D> {
+        let changeset = self.chain.insert_block(block_id)?;
+        let changed = !changeset.is_empty();
+        if changed {
+            self.persist.stage(changeset.into());
```
another redundant branch.

LLFourn reviewed (changes requested):
Design looks good. I think we should try and avoid adding another argument to `insert_tx` on Wallet.

evanlinjin committed:
Add `last_seen` to the the `ConfirmationTime::Unconfirmed` variant

This allows us to skip adding an extra input to `Wallet::insert_tx`.

Also remove redundant logic.

evanlinjin committed:
[electrum_redesign] Introduce redesigned `ElectrumExt`

There are a number of improvements that can be done, but it is in a
decent state to be usable.

Possible improvements:

* Remove requirement to retry obtaining ALL data after reorg is
  detected. Transactions can be anchored to a lower block (not block
  tip), and an `assume_final_depth` value can be used.

* The logic to finalize an update with confirmation time can be improved
  during reorgs to not require returning an error.

evanlinjin committed:
[examples_redesign] Introduce `example_cli` package

This is the equivalent of `keychain_tracker_example_cli` that works with
the redesigned structures.

evanlinjin committed:
[examples_redesign] Implemented `example_electrum`

This is a version of `keychain_tracker_electrum` that uses the
redesigned structures instead.

LLFourn commented on example-crates/example_cli/src/lib.rs:
```diff
@@ -0,0 +1,780 @@
+where
+    ChangeSet<A, X>: Default + Append + DeserializeOwned + Serialize,
+{
+    let process_spk = |spk_i: u32, spk: &Script, index_additions: DerivationAdditions<Keychain>| {
```
Make what is happening in this closure so easy to do it doesn't need a closure.

For example, make a few `From` implementations or constructor functions easily create the `ChangeSet` in one line.
There is no need for the branch on `is_empty`.

evanlinjin replied:
Already changed in 6f5a2cc56e95b1fb16a88590cf2698a7a75c1209

LLFourn committed:
[example-cli] simplify new address logic

evanlinjin committed:
Introduce `keychain::LocalChangeSet`

This corresponds to `keychain::KeychainChangeSet` but for the redesigned
structures with `LocalChain`.

This structure is now used in `Wallet` as well as the examples.

LLFourn commented on crates/electrum/src/v2.rs:
```diff
@@ -0,0 +1,507 @@
+                let _ = graph_update.insert_anchor(txid, anchor);
+            }
+        }
+        dbg!(graph_update.full_txs().count());
```
```suggestion
```
I'm guess this wasn't meant to be here.

LLFourn commented on crates/electrum/src/v2.rs:
```diff
@@ -0,0 +1,507 @@
+                "tip changed during update: update_tip={:?}",
+                update_tip
+            )));
+        }
```
I think we should just note that the confirmation time might not be precisely correct if there was a re-org. Throwing an error here is worse than returning slightly off times.

I think electrum really intends for you to use the merkle proof API. Best solution is to do that and attach confirmation time of each block to `LocalChain`.

evanlinjin replied:
I think so too, but probably for a separate PR?

evanlinjin replied:
Issue created: https://github.com/bitcoindevkit/bdk/issues/980

LLFourn commented on crates/electrum/src/v2.rs:
```diff
@@ -0,0 +1,507 @@
+    txid: Txid,
+) -> Option<ConfirmationHeightAnchor> {
+    if txid
+        == Txid::from_hex("4a5e1e4baab89f3a32518a88c31bc87f618f76673e2cc77ab2127b7afdeda33b")
```
this deserves a comment why it's necessary.

LLFourn commented on example-crates/example_cli/src/lib.rs:
```diff
@@ -0,0 +1,728 @@
+            // If we're unable to persist this, then we don't want to broadcast.
+            db.lock()
+                .unwrap()
+                .stage(C::from(KeychainAdditions::from(index_additions)));
```
The comment says we're going to try and persist this to disk but then it just stages it.

evanlinjin replied:
Changed to this:

```rust
            // We must first persist to disk the fact that we've got a new address from the
            // change keychain so future scans will find the tx we're about to broadcast.
            // If we're unable to persist this, then we don't want to broadcast.
            {
                let db = &mut *db.lock().unwrap();
                db.stage(C::from(KeychainAdditions::from(index_additions)));
                db.commit()?;
            }
```

LLFourn commented on example-crates/example_cli/src/lib.rs:
```diff
@@ -0,0 +1,728 @@
+    }
+}
+
+pub fn prepare_index<S: clap::Subcommand, CTX: secp256k1::Signing>(
```
Does this need to be `pub`. Does it need to exist at all? I think you can just inline it.

evanlinjin replied:
Inlined.

LLFourn reviewed (changes requested):
The examples are nice.

Instead of commenting out all these files can you just remove them from workspace for a while?

danielabrozzoni commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -497,41 +468,70 @@ impl<D> Wallet<D> {
         Ok(changed)
     }
 
     #[deprecated(note = "use Wallet::transactions instead")]
```
Can't we just remove this whole method?

evanlinjin replied:
Removed.

danielabrozzoni commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -1685,52 +1689,62 @@ impl<D> Wallet<D> {
     ///
     /// [`commit`]: Self::commit
-    pub fn apply_update(&mut self, update: Update) -> Result<(), UpdateError>
+    pub fn apply_update(&mut self, update: Update) -> Result<bool, UpdateNotConnectedError>
```
nit: the doc should explain the return type

evanlinjin replied:
Added.

danielabrozzoni commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -1685,52 +1689,62 @@ impl<D> Wallet<D> {
     ///
     /// [`staged`]: Self::staged
-    pub fn commit(&mut self) -> Result<(), D::WriteError>
+    pub fn commit(&mut self) -> Result<bool, D::WriteError>
```
ditto above

evanlinjin replied:
Done.

danielabrozzoni commented on crates/electrum/src/v2.rs:
```diff
@@ -0,0 +1,507 @@
+            keychain_update: Default::default(),
+        }
+    }
+}
```
I think this can be derived

evanlinjin replied:
The problem with using `derive(Default)` is that `K` and `A` will also be forced to implement `Default` (which are unnecessary bounds).

danielabrozzoni commented on example-crates/example_cli/Cargo.toml:
```diff
@@ -0,0 +1,17 @@
+[package]
+name = "example_cli"
```
Am I right in saying that `example_cli` will eventually replace `keychain_tracker_example_cli`?

evanlinjin replied:
Yes

danielabrozzoni commented on example-crates/example_cli/src/lib.rs:
```diff
@@ -0,0 +1,728 @@
+    Send {
+        value: u64,
+        address: Address,
+        #[clap(short, default_value = "largest-first")]
```
I would use bnb by default

evanlinjin replied:
Changed.

danielabrozzoni commented on example-crates/example_cli/src/lib.rs:
```diff
@@ -0,0 +1,728 @@
+pub enum TxOutCmd {
+    List {
+        /// Return only spent outputs.
+        #[clap(short, long)]
+        spent: bool,
+        /// Return only unspent outputs.
+        #[clap(short, long)]
+        unspent: bool,
```
nit: This could probably be an enum spent_only/unspent_only/all. Same for confirmed.

evanlinjin replied:
This idea deserves to be addressed in a separate ticket/PR!

danielabrozzoni commented on example-crates/example_cli/src/lib.rs:
```diff
@@ -0,0 +1,728 @@
+
+    let mut transaction = Transaction {
+        version: 0x02,
+        lock_time: chain
+            .get_chain_tip()?
+            .and_then(|block_id| LockTime::from_height(block_id.height).ok())
+            .unwrap_or(LockTime::ZERO)
+            .into(),
```
This works because we never add timelocks to the assets, so they won't be used, and the lock_time can be the current block for fee sniping purposes.

...But it took me a sec to figure it out, so maybe can you add a small comment?

danielabrozzoni commented on example-crates/example_electrum/src/main.rs:
```diff
@@ -0,0 +1,323 @@
+
+#[derive(Subcommand, Debug, Clone)]
+enum ElectrumCommands {
+    /// Scans the addresses in the wallet using the esplora API.
```
nit: electrum, not esplora :)

danielabrozzoni commented on example-crates/example_electrum/src/main.rs:
```diff
@@ -0,0 +1,323 @@
+        #[clap(flatten)]
+        scan_options: ScanOptions,
+    },
+    /// Scans particular addresses using the esplora API.
```
ditto above

danielabrozzoni commented on example-crates/example_electrum/src/main.rs:
```diff
@@ -0,0 +1,323 @@
+            stop_gap,
+            scan_options,
+        } => {
+            let (keychain_spks, c) = {
```
Can you give a more explanatory name to `c`?

evanlinjin replied:
`local_chain`.

danielabrozzoni reviewed (commented):
A few comments, I'm still not done reviewing electrum v2. I agree with lloyd that the old ~~electrum~~ example files should be deleted :)

evanlinjin committed:
Various tweaks to code arrangement and documentation

As per reviews by @danielabrozzoni and @LLFourn

evanlinjin commented:
Oh yeah I still need to remove the commented-out examples.

evanlinjin committed:
`bdk_electrum` API improvements and simplifications

* `ElectrumUpdate::missing_full_txs` now returns a `Vec<Txid>` so we
  don't keep a reference to the passed-in `graph`.

* `ElectrumUpdate::finalize*` methods now takes in `missing` txids
  instead of `full_txs`. `Client::batch_transaction_get` is called
within the methods.

Other changes:

* `wallet::ChangeSet` is now made public externally. This is required as
  a wallet db should implement `PersistBackend<wallet::ChangeSet>`.

evanlinjin committed:
Update the `wallet_electrum` example

evanlinjin committed:
Implement redesigned versions of `EsploraExt` and `EsploraAsyncExt`

All associated examples are also updated.

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -497,41 +469,57 @@ impl<D> Wallet<D> {
+                }
+                (
+                    Some(ConfirmationTimeAnchor {
+                        anchor_block: tip.expect("already checked if tip_height > height"),
```
shouldn't this be the tightest block. i.e. the lowest hight in local chain that is `>=` the height.

evanlinjin replied:
Yes, changed in 9653b06ff06a3ae605393256b45b0c6685230291

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -497,41 +469,57 @@ impl<D> Wallet<D> {
-    ) -> Result<bool, chain_graph::InsertTxError<ConfirmationTime>> {
-        let changeset = self.keychain_tracker.insert_tx(tx, position)?;
-        let changed = changeset.is_empty();
+    ) -> Result<bool, InsertTxError>
```
I actually hadn't considered the implications of this change to wallet. I think what you've done makes sense.

evanlinjin replied:
Added an additional comment in 9653b06ff06a3ae605393256b45b0c6685230291

evanlinjin committed:
[chain_redesign] Remove old structures

Other changes:

* The `async-https` feature of `bdk_esplora` is no longer default.
* Rename `ObservedAs` to `ChainPosition`.
* Set temporary MSRV to 1.60.0 to compile all workspace members will all
  features.

vladimirfomene commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -249,15 +263,15 @@ impl<D> Wallet<D> {
 
     /// Iterator over all keychains in this wallet
     pub fn keychanins(&self) -> &BTreeMap<KeychainKind, ExtendedDescriptor> {
```
Nits: This should be `keychains`

vladimirfomene commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -472,10 +441,13 @@ impl<D> Wallet<D> {
+    {
+        let changeset = self.chain.insert_block(block_id)?;
+        let changed = !changeset.is_empty();
+        self.persist.stage(changeset.into());
```
Correct me if I'm wrong but shouldn't we only persist if `changed` is true?

evanlinjin replied:
That was my intuition originally as well. However, @LLFourn suggested that the branch is not needed as `stage()` wouldn't do anything with an empty changeset anyway.

vladimirfomene commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -497,41 +469,57 @@ impl<D> Wallet<D> {
+
+        let changeset: ChangeSet = self.indexed_graph.insert_tx(&tx, anchor, last_seen).into();
+        let changed = !changeset.is_empty();
         self.persist.stage(changeset);
```
Again is there a need to stage these changes if `changed` is false.

vladimirfomene commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -1019,16 +1003,21 @@ impl<D> Wallet<D> {
+
+        let pos = graph
+            .get_chain_position(&self.chain, chain_tip, txid)
+            .ok_or(Error::TransactionNotFound)?;
```
Maybe we need a better error variant here. For saying the associated chain position for the txid was not found.

evanlinjin replied:
I agree that the errors aren't great.

However, I disagree about having a single error enum with a large number of variants. I think we should fix errors in one go with a separate PR, otherwise it's duplicate work.

evanlinjin replied:
Created a separate issue.

vladimirfomene commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -1681,56 +1673,71 @@ impl<D> Wallet<D> {
+        changeset.append(self.indexed_graph.apply_update(update.graph).into());
+
+        let changed = !changeset.is_empty();
         self.persist.stage(changeset);
```
As above, I believe we should only stage changes if `changed` is false.

vladimirfomene reviewed (commented):
Great work Evan! Just have a couple of comments

LLFourn commented on crates/chain/src/chain_data.rs:
```diff
@@ -99,44 +47,31 @@ impl TxHeight {
 pub enum ConfirmationTime {
     Confirmed { height: u32, time: u64 },
-    Unconfirmed,
+    Unconfirmed { last_seen: u64 },
```
Can we document each variant. Specify the units of last_seen and time.

It's time to add `#[warn(missing_docs)]` on this project and make sure CI is failing on it. Fine to do it in a followup PR but i think it should be done before making the next alpha release.

evanlinjin replied:
Documentation and `#[warn(missing_docs)]` added in 250458ae01e7157c97f3c88f572e4d0ff21c019f

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -83,19 +86,18 @@ const COINBASE_MATURITY: u32 = 100;
-    persist: persist::Persist<KeychainKind, ConfirmationTime, D>,
+    chain: LocalChain,
+    indexed_graph: IndexedTxGraph<ConfirmationTimeAnchor, KeychainTxOutIndex<KeychainKind>>,
+    persist: Persist<D, ChangeSet>, // [TODO] Use a different `ChangeSet`
```
What does this TODO mean? Can you remove any TODOs you've made during this PR and make an issue for it and reference the issue instead of just `//TODO` in code.

evanlinjin replied:
This TODO is not needed, there is nothing to TODO here. Removed.

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -271,23 +285,24 @@ impl<D> Wallet<D> {
-                self.persist.stage(changeset.into());
+                self.persist
+                    .stage(ChangeSet::from(IndexedAdditions::from(index_additions)));
                 self.persist.commit().expect("TODO");
```
Created https://github.com/bitcoindevkit/bdk/issues/996

evanlinjin replied:
Partially resolved in 250458ae01e7157c97f3c88f572e4d0ff21c019f

I think we should get rid of the address helper methods and just have one method that uses `KeychainKind` as a second input.

LLFourn reviewed (approved):
Fantastic work. ACK 965cbe6dde99a1a58168cc0f4bdd320f0aba10e1.

Left nits and notes.

LLFourn commented on example-crates/example_electrum/src/main.rs:
```diff
@@ -106,29 +141,40 @@ fn main() -> anyhow::Result<()> {
+                    .rev()
+                    .take(ASSUME_FINAL_DEPTH)
+                    .map(|(k, v)| (*k, *v))
+                    .collect::<BTreeMap<u32, BlockHash>>();
```
Made issue about the ergonomics here: https://github.com/bitcoindevkit/bdk/issues/997

evanlinjin committed:
[wallet_redesign] Modified `insert_tx` to use lowest checkpoint

Also updated the documentation.

evanlinjin committed:
[wallet_redesign] Clean up and document address methods

evanlinjin committed:
Update documentation

* Add `warn(missing_docs)` for `bdk_wallet` and `bdk_chain`.
* Add missing documentation.
* Remove `LocalChain::heights` method.
* Remove old TODOs.

danielabrozzoni commented on crates/file_store/README.md:
```diff
@@ -1,9 +1,9 @@
+[`Persist`](`bdk_chain::Persist`).
 
-The main structure is [`KeychainStore`](`crate::KeychainStore`), which can be used with [`bdk`]'s
+The main structure is [`Store`](`crate::Store`), which can be used with [`bdk`]'s
```
Do you think we will eventually build some convenience structs on top of `Store`, similar to what `KeychainStore` was, or will users just use `Store`?

evanlinjin replied:
What convenience structs existed on top of `KeychainStore`? What do you think is missing with `Store`?

danielabrozzoni commented on crates/chain/src/chain_data.rs:
```diff
@@ -98,45 +48,44 @@ impl TxHeight {
-        }
+impl ConfirmationTime {
+    /// Construct an unconfirmed variant using the given `last_seen` time in unix seconds.
+    pub fn unconfirmed(last_seen: u64) -> Self {
```
Is there a reason for not having also the `confirmed` constructor? 
I'd say either have both or none of them, the asymmetry here bothers me 😅

evanlinjin replied:
No particular reason, I only created the `unconfirmed` constructor because it became tedious.

danielabrozzoni replied:
Then I would add the `confirmed` as well

danielabrozzoni commented on crates/chain/src/chain_data.rs:
```diff
@@ -235,75 +185,32 @@ impl Anchor for ConfirmationTimeAnchor {
     pub txout: TxOut,
     /// The position of the transaction in `outpoint` in the overall chain.
-    pub chain_position: P,
+    pub chain_position: ChainPosition<A>,
```
I think this makes things way clearer btw :)

evanlinjin replied:
This was only possible after getting rid of the old structures.

danielabrozzoni commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -203,6 +208,24 @@ impl<A: Anchor, IA: Append> Append for IndexedAdditions<A, IA> {
+impl<A, K> From<DerivationAdditions<K>> for IndexedAdditions<A, DerivationAdditions<K>> {
+    fn from(index_additions: DerivationAdditions<K>) -> Self {
+        Self {
+            graph_additions: Default::default(),
```
tiny tiny nit but if you're using `..Default::default()` a couple of lines above I think you should use it here too

evanlinjin replied:
Sorry @danielabrozzoni, I'm going to ignore this one :sweat_smile:

danielabrozzoni replied:
🤷🏻‍♀️

danielabrozzoni commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -271,91 +291,99 @@ impl<D> Wallet<D> {
-    fn _get_address(&mut self, address_index: AddressIndex, keychain: KeychainKind) -> AddressInfo
+    /// Return a derived address using the specified `keychain` (external/internal).
+    ///
+    /// If `keychain` is [`KeychainKind::External`], external addresses will be derived (used for
```
nit: it might be easier to read written as "If `keychain` is [`KeychainKind::External`], external addresses (used for receiving funds) will be derived."

It might be just me though

(Ditto below w/ internal addresses)

danielabrozzoni commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -271,91 +291,99 @@ impl<D> Wallet<D> {
+    /// receiving funds).
+    ///
+    /// If `keychain` is [`KeychainKind::Internal`], internal addresses will be derived (used for
+    /// creating change outputs). If the wallet does not have an internal keychain, it will use the
```
Mhh I would expect an error if the user asks for an internal address without having the internal keychain

notmandatory replied:
This is how the current BDK API works, instead of changing in this PR can we make an issue and discuss there?

danielabrozzoni commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -472,66 +460,91 @@ impl<D> Wallet<D> {
+            &self.chain,
+            self.chain.tip().unwrap_or_default(),
+            self.indexed_graph.index.outpoints().iter().cloned(),
+            |&(k, _), _| k == KeychainKind::Internal,
```
This reminds me that I had a nice chat w/ some friends regarding the meaning of "trusted" and "untrusted" keychain and how it doesn't really make sense as a distinction. I will post a message on discord explaining (eventually 😅 ) or directly update the balance issue we already have 

But the TL;DW is: addresses are public anyways, everyone can send money to our change addresses, so you should distinguish between trusted and untrusted based on whether you did a certain tx or not, not based on the keychain used

notmandatory replied:
I added a note on #797 so we don't forget to continue the discussion there.

danielabrozzoni reviewed (commented):
Partial review for now because I'm too tired to finish it, I looked at `bdk`, `chain` and `file_store`, they look good to me, I just have a few questions/nits

danielabrozzoni commented on crates/esplora/README.md:
```diff
@@ -27,7 +27,7 @@ To use the extension traits:
 use bdk_esplora::EsploraExt;
 // for async
-use bdk_esplora::EsploraAsyncExt;
+// use bdk_esplora::EsploraAsyncExt;
```
Why can't it be uncommented? Doesn't compile?

danielabrozzoni commented on crates/esplora/Cargo.toml:
```diff
@@ -13,12 +13,12 @@ readme = "README.md"
 
 [features]
-default = ["async-https", "blocking"]
+default = ["blocking"]
```
Why isn't async-https included by default anymore? Is it incompatible with blocking?

notmandatory replied:
For what it's worth, I tried putting `"async-https"` and the below `use bdk_esplora:EsploraAsyncExt;` back in and all tests build and run fine.

evanlinjin replied:
I think this was an attempt to keep MSRV to 1.57 after the `base64` mishap that is now fixed. Will change it back.

danielabrozzoni commented on crates/esplora/src/async_ext.rs:
```diff
@@ -100,47 +92,35 @@ impl EsploraAsyncExt for esplora_client::AsyncClient {
+                    .insert_block(update_block_id)
+                    .expect("cannot repeat height here");
+                if update_block_id.hash == original_hash {
+                    break;
+                }
+            }
 
-        for (&height, &original_hash) in local_chain.iter().rev() {
-            let update_block_id = BlockId {
-                height,
-                hash: self.get_block_hash(height).await?,
+            let tip_at_start = BlockId {
+                height: self.get_height().await?,
+                hash: self.get_tip_hash().await?,
             };
-            let _ = update
-                .insert_checkpoint(update_block_id)
-                .expect("cannot repeat height here");
-            if update_block_id.hash == original_hash {
-                break;
+
+            if update.chain.insert_block(tip_at_start).is_ok() {
+                break (update, tip_at_start);
             }
-        }
-        let tip_at_start = BlockId {
-            height: self.get_height().await?,
-            hash: self.get_tip_hash().await?,
         };
```
I really need some more comments in order to understand what's happening here 😅  To me it seems that you're trying to build an update to the local chain that doesn't have any block that might have been reorged:
1 You traverse the local chain backwards
2 For each block height, you retrieve the current block hash
3 You insert the block height and new hash in the update
4 If the hash is the same as the one you already had in the local chain, then you break as you removed any possible reorg
5. You insert the tip, if it's all good then you're done, otherwise you start from scratch as it means you were updating it mid-reorg, and it's better to start again

Does this sound fair? If so, can you copy-paste these steps in the code as comments? 
(Also, do you think it would work if you swapped step 3 and 4? To me it seems more logical to insert in the update only blocks that actually changed, but i don't know how the localchain works, and it might need at least one block in common to be able to update...)

evanlinjin replied:
I'll update the comments to try make thinks clearer. However, by principle, I don't think comments should say what is happening (because you can just read the code). The "why" is more important.

danielabrozzoni commented on crates/esplora/src/async_ext.rs:
```diff
@@ -149,7 +129,7 @@ impl EsploraAsyncExt for esplora_client::AsyncClient {
 
             loop {
-                let futures: FuturesOrdered<_> = (0..parallel_requests)
+                let futures = (0..parallel_requests)
```
I know you haven't modified this, but I think we can make the code below a bit clearer. Copy-pasting it here as I can't comment on this PR (it's too far away from modified stuff). Lines 143 to 158:
```
                            if n_confirmed >= 25 {
                                loop {
                                    let new_related_txs = client
                                        .scripthash_txs(
                                            &script,
                                            Some(related_txs.last().unwrap().txid),
                                        )
                                        .await?;
                                    let n = new_related_txs.len();
                                    related_txs.extend(new_related_txs);
                                    // we've reached the end
                                    if n < 25 {
                                        break;
                                    }
                                }
                            }

```

I think we could substitute `if n_confirmed >= 25 { loop { ` with a while loop (so that we don't even need the break on like 155). Also I think we should say "n_confirmed == 25" since if I understand what's happening it can  never be more than 25

evanlinjin replied:
> it can never be more than 25

I don't think this statement is correct.
https://github.com/Blockstream/esplora/blob/master/API.md#get-scripthashhashtxs

I think we should leave it as is. These are style nits.

danielabrozzoni commented on crates/esplora/src/async_ext.rs:
```diff
@@ -180,36 +160,23 @@ impl EsploraAsyncExt for esplora_client::AsyncClient {
                     if related_txs.is_empty() {
                         empty_scripts += 1;
                     } else {
                         last_active_index = Some(index);
```
Waaaaait are the responses in `idx_with_tx` ordered? Shouldn't we do `last_active_index = max(last_active_index, index);`?

evanlinjin replied:
Yes, the responses are in order. Did you miss the fact that we are collecting into a `FuturesOrdered`?

danielabrozzoni commented on crates/esplora/src/async_ext.rs:
```diff
@@ -267,50 +235,35 @@ impl EsploraAsyncExt for esplora_client::AsyncClient {
-            // A reorg occurred, so let's find out where all the txids we found are in the chain now.
-            // XXX: collect required because of weird type naming issues
+        if tip_at_start.hash != self.get_block_hash(tip_at_start.height).await? {
+            // A reorg occurred, so let's find out where all the txids we found are now in the chain
```
It seems to me that this piece of code is important to find out what happened after a reorg, but we do it only if the reorg happens during the scan, and not every time there's a reorg. Am I wrong?

evanlinjin replied:
If a reorg happens during our scan, our view of blocks may not be consistent so we need to retry. We just need to ensure our update is consistent. If our update connects with our original view, we can assume that the resultant state is also consistent.

danielabrozzoni commented on example-crates/wallet_esplora/Cargo.toml:
```diff
@@ -1,5 +1,5 @@
 [package]
-name = "bdk-esplora-wallet-example"
+name = "wallet_esplora"
```
A C K ✨

danielabrozzoni reviewed (commented):
I ended up reviewing all the esplora code, and now I'm full of questions, even regarding stuff that wasn't touched in this PR. Sorry haha

danielabrozzoni commented on example-crates/example_cli/src/lib.rs:
```diff
@@ -0,0 +1,736 @@
+    pub command: Commands<S>,
+}
+
+#[allow(clippy::almost_swapped)]
```
Weird, I don't have warnings for clippy on cargo 1.69 nor 1.57 🤷🏻‍♀️

evanlinjin replied:
I think it was caused by this regression (https://github.com/rust-lang/rust-clippy/issues/10421) which is fixed in the latest version for `cargo clippy`. Will remove.

notmandatory commented:
I finally tested this branch using a signet descriptor setup with a lot of addresses to scan and a slightly modified version of `example-crates/wallet_esplora_async`. My test descriptor has 1001 UTXO (500 sats each) spread over 200_000 addresses. Good news is it sync'd without a problem 🎉 . Bad news is it did take a bit of time. I'll retest with more parallel threads and see if that helps. 

cc @bodymindarts 

Test Params:
```rust
const STOP_GAP: usize = 250;
const PARALLEL_REQUESTS: usize = 500;
const ESPLORA_API_URL: &str = "https://mempool.space/signet/api";
const EXTERNAL_DESC: &str = "wpkh([f6c175d5/84'/1'/0']tpubDD3t7WfvBLbiVB1BbxZjeusz76Xb1Hdv1AgjdyEuVWHzbBNN1Fio9X5rfUpk6JWyuTM7YHLQ7CsetUYMtWm3USAFaHzYY4ymd1FXfum76p5/0/*)#2xxpq0d7";
const INTERNAL_DESC: &str = "wpkh([f6c175d5/84'/1'/0']tpubDD3t7WfvBLbiVB1BbxZjeusz76Xb1Hdv1AgjdyEuVWHzbBNN1Fio9X5rfUpk6JWyuTM7YHLQ7CsetUYMtWm3USAFaHzYY4ymd1FXfum76p5/1/*)#mjrqa6ax";
```

Test Results:

```
Wallet balance after syncing: 500500 sats
Wallet sync took 16 minutes and 31 seconds
```
DB file size: 26K

PS. I've been unable to get this test descriptor to sync with bdk 0.28, still trying to nailing down the issue. 

EDIT1:

With `PARALLEL_REQUESTS: usize = 1000;` the sync time was:
```
Wallet sync took 9 minutes and 4 seconds
```
The sync time should scale well with more threads as long as the esplora server is properly sized.

EDIT2: 

With `PARALLEL_REQUESTS` at 2000:
```
Wallet sync took 6 minutes and 21 seconds
```

notmandatory reviewed (commented):
This is a monumental piece of work. I like all the new type naming. Also it's much easier to read with old code removed.  At our team call we'll discuss where it needs more detailed review and how to do it. If everyone is OK with the `chain`, `electrum` and `esplora` crates I think anything else can be fixed in future alpha releases.

danielabrozzoni commented on example-crates/example_cli/src/lib.rs:
```diff
@@ -0,0 +1,736 @@
+pub use anyhow;
```
This whole file would be way easier to review if there was one move-only commit, followed by commits changing stuff, if needed. In this way, I could just review the commits changing the logic. Just something to keep in mind for next time!

evanlinjin replied:
Apologies. Will keep in mind for next time.

danielabrozzoni commented on example-crates/example_electrum/src/main.rs:
```diff
@@ -192,54 +242,77 @@ fn main() -> anyhow::Result<()> {
+                .blocks()
+                .iter()
+                .rev()
+                .take(ASSUME_FINAL_DEPTH)
```
I'm a bit lost. It seems that you're taking the latest ASSUME_FINAL_DEPTH blocks. Why? Can you add a comment in the code explaining what's happening?

evanlinjin replied:
Will add a comment. Basically anything as deep or deeper than `ASSUME_FINAL_DEPTH` will not be reorged out (based on our assumption). This is essentially the minimum checkpoints from the tip to include (from the original chain) to guarantee that we will find a point of agreement.

LLFourn replied:
Note that #1002 is removing this ASSUME_FINAL_DEPTH thing.

danielabrozzoni commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -0,0 +1,486 @@
+            .collect()
+    }
+
+    pub fn finalize(
```
A comment explaining what it means to finalize an update would be helpful

danielabrozzoni commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -0,0 +1,486 @@
+
+impl ElectrumExt<ConfirmationHeightAnchor> for Client {
+    fn get_tip(&self) -> Result<(u32, BlockHash), Error> {
+        // TODO: unsubscribe when added to the client, or is there a better call to use here?
```
What does it mean to "unsubscribe when added to the client"?

evanlinjin replied:
The electrum API has a subscription model. We are subscribing to new tips coming in (we get the first tip), but we don't process later tips, so we should unsubscribe from receiving more tips.

danielabrozzoni replied:
> The electrum API has a subscription model. We are subscribing to new tips coming in (we get the first tip), but we don't process later tips, so we should unsubscribe from receiving more tips.

Yep, but why "when added to the client" - unsubscribe when what is added to the client, the first tip?

danielabrozzoni commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -0,0 +1,486 @@
+    }
+}
+
+fn populate_with_outpoints<K>(
```
Pls put a comment explaining what this method does. What is being populate with outpoints? The update?

evanlinjin replied:
Yes, the update. We already have a comment on `scan` explaining the `outpoints` input though?

danielabrozzoni commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -0,0 +1,486 @@
+            None => continue,
+        };
+        // attempt to find the following transactions (alongside their chain positions), and
+        // add to our sparsechain `update`:
```
not a sparsechain anymore :)

danielabrozzoni commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -0,0 +1,486 @@
+        }
+    }
+
+    // Insert the new tip so new transactions will be accepted into the sparsechain.
```
Not a sparsechain anymore

danielabrozzoni commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -0,0 +1,486 @@
+            .map(|data| (data.height as u32, data.header.block_hash()))
+    }
+
+    fn scan<K: Ord + Clone>(
```
I think this whole method needs comments explaining what's going on

danielabrozzoni reviewed (commented):
I finished reviewing electrum as well, I haven't really looked into electrum_ext as I had troubles figuring out what was happening, I left a few comments asking for more comments (lol)

It's probably obvious at this point but I really think it would have helped w/ reviewing and updating if this whole PR was split in at least 3 or 4 smaller ones (wallet, esplora, electrum, example cli). Something to keep in mind for next time!

LLFourn reviewed (approved):
ACK 75f8b81d58a985669ce7302fe235ad68eddc0d47

I think the way to resolve the remaining discussion here is to make issues for them if the concerns still exist after #1002

danielabrozzoni reviewed (approved):
Partial ACK 75f8b81d58a985669ce7302fe235ad68eddc0d47 - the Wallet code looks good to me, I don't have a good enough understanding of the esplora/electrum code to confidently ACK it.
